refactor(stdlib): rewrite `contractAddressExt`, `newAddress` and `Address.asSlice` functions to Tact; add `asAddressUnsafe` and `contractHash`

i582 opened the pull request:
<img width="875" alt="Screenshot 2025-02-13 at 16 00 12" src="https://github.com/user-attachments/assets/8f1cc9cb-130a-4f3c-81d4-cccb7f66afb0" />

## Issue

Closes #1763.
Closes #1764.
Closes #1765.
Towards #1736.
Towards #573.

## Checklist

- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

i582 committed:
refactor(stdlib: rewrite `contractAddressExt`, `newAddress` and `Address.asSlice` functions to Tact

Fixes #1763
Fixes #1764
Fixes #1765

i582 committed:
optimize

i582 committed:
update stdlib

i582 committed:
update snapshots

i582 committed:
fixes

i582 committed:
revert optimization for now

Shvandre commented on src/stdlib/stdlib/std/internal/contract.tact:
```diff
@@ -1,21 +1,38 @@
+}
+
+asm extends fun toAddressUnsafe(self: Slice): Address { NOP }
+asm extends fun asSlice(self: Address): Slice { NOP }
```
Idk, does those NOPs presented in resulting code? Maybe we should define those by just curly braces, without NOP? (NOP actually consumes gas. Insignificant amount, but still consumes)

i582 replied:
Looks like yes:
<img width="262" alt="Screenshot 2025-02-13 at 14 42 38" src="https://github.com/user-attachments/assets/b9e08b9c-30bf-4ec2-bb00-cfc78ec2c0d2" />

But currently, empty assembly functions result in the following FunC compilation error:

```
FunC compilation error: /Users/petrmakhnev/tact/src/test/benchmarks/contracts/output/address_Addresses.code.fc:51:4: error: string with assembler instruction expected
  """;
```

Let's create and issue for such cases

Commit:
Merge branch 'main' into pmakhnev/rewrite-address-related-functions-to-tact

i582 committed:
fix jetton benchmarks

i582 committed:
add benchmark

i582 commented:
Some changes from #1804, let's wait for #1804 to be merged and merge master into this PR

i582 committed:
Merge branch 'main' into pmakhnev/rewrite-address-related-functions-to-tact

i582 committed:
revert some changes

Shvandre commented on src/stdlib/stdlib/std/internal/contract.tact:
```diff
@@ -1,21 +1,36 @@
+asm extends fun toAddressUnsafe(self: Slice): Address { NOP }
+asm extends fun asSlice(self: Address): Slice { NOP }
+
+inline fun contractAddressExt(chain: Int, code: Cell, data: Cell): Address {
```
Lets optimize this function. It shouldn't use endCell().
Let's use HASHEXT_SHA256 to calculate hash of builder

i582 replied:
Done

i582 committed:
optimize `contractAddressExt` with custom `calculateAddressHash` instead of `endCell().hash()`

i582 committed:
fix

novusnota commented on src/stdlib/stdlib/std/internal/contract.tact:
```diff
@@ -1,21 +1,60 @@
+    return builderSha256(builder);
+}
+
+asm fun builderSha256(b: Builder): Int {
```
Do we need this as a new stdlib library function? Perhaps let's use the approach from: https://github.com/tact-lang/tact/pull/1626

novusnota replied:
Besides, the `.toAddressUnsafe()` and `calculateAddressHash()` are introduced too. If we want to keep them in the stdlib, let's add an entry in the CHANGELOG.

If not, let's use the approach from https://github.com/tact-lang/tact/pull/1626

i582 replied:
Let's not add any extra FunC functions as we want to remove them for the native backend

novusnota replied:
Ok, then what about naming `builderSha256()` like `sha256Builder()`? Or, even better, re-using the on-chain `sha256()` we have, but with converting the Builder to a Slice?

i582 replied:
>  but with converting the Builder to a Slice?

It's too expensive because of `ENDC`

i582 replied:
Renamed to `Builder.hash()`

i582 committed:
added entry to CHANGELOG.md

i582 committed:
builderSha256 -> Builder.hash

anton-trunov commented:
merge conflicts

i582 committed:
Merge branch 'main' into pmakhnev/rewrite-address-related-functions-to-tact

i582 committed:
update results

i582 committed:
fix

i582 committed:
fix

i582 commented:
> merge conflicts

Fixed

anton-trunov commented on src/stdlib/stdlib/std/internal/cells.tact:
```diff
@@ -163,6 +163,11 @@ asm extends fun refs(self: Builder): Int { BREFS }
 
 asm extends fun bits(self: Builder): Int { BBITS }
 
+asm extends fun hash(self: Builder): Int {
```
we need correctness tests for this new method: the hash we compute here should have a certain relationship with `.hash()` for Cells / Slices

Gusarich replied:
This one will calculate data hash, not cell hash. Exactly like `SHA256U` for slices.

I think we should either rename this method to something like `dataHash` or make it compute cell hash to match the logic of `.hash` for cells and slices.

novusnota replied:
:eyes: https://github.com/tact-lang/tact/pull/1766#issuecomment-2658464665

novusnota committed:
docs: doc comments for the rest of contract.tact and adjustments in the public docs

novusnota committed:
docs/chore: rename `calculateAddressHash` to `contractHash`

novusnota committed:
Merge branch 'pmakhnev/rewrite-address-related-functions-to-tact' of github.com:tact-lang/tact into pmakhnev/rewrite-address-related-functions-to-tact

novusnota committed:
Merge branch 'main' into pmakhnev/rewrite-address-related-functions-to-tact

novusnota commented:
Added docs, please take a look!

Renamed `calculateAddressHash()` to `contractHash()` for clarity and conformance to existing functions. It's not the `Address` yet only the `account_id` part of the address, hence the simplification of the name.

Also renamed `Slice.toAddressUnsafe` to `Slice.asAddressUnsafe` because:
* there is `map.asCell()` method that does similar type casting in spirit, i.e. NOP
* there is a mirror function `Address.asSlice()`, also a NOP
* issue 1736 originally discussed the `.asAddressUnsafe()` name

In general, I think at some point we might carefully split all the `asSomething()` casts and `toSomething()` conversions, where the former are no-ops and simply instruct the compiler to start viewing the underlying type differently, while the latter actually require some operations and transformations to be performed.

P.S.: We can revert those renames if you wish, but I believe they are on point :)

novusnota commented:
_I've renamed `calculateAddressHash()` to `contractHash()`, see the rationale [here](https://github.com/tact-lang/tact/pull/1766#issuecomment-2658464256)._

The current implementation of `contractHash()` produces the following Fift assembly. Notice lots of `SWAP`s and such:

```ts
$global_contractHash PROCINLINE:<{
      NEWC
  131380 PUSHINT
  SWAP
  24 STU
  s2 PUSH
      CDEPTH
  SWAP
  16 STU
  OVER
      CDEPTH
  SWAP
  16 STU
  s0 s2 XCHG
      HASHCU
  ROT
  256 STU
  SWAP
      HASHCU
  SWAP
  256 STU
      ONE HASHEXT_SHA256
}>
```

If we rewrite it as a Tact `asm` function, we can easily get rid of lots of `SWAP`s and similar instructions by placing the data optimally, thus optimizing the resulting gas consumption. Furthermore, we would not need to introduce a separate `Builder.hash()` function at all, since it would be inlined.

That's why instead of writing docs for `Builder.hash()` function, I've efficiently rewrote the `contractHash()` function to Tact assembly:

```tact
asm fun contractHash(code: Cell, data: Cell): Int {
    // According to the https://docs.tact-lang.org/book/cells#cells-representation,
    // the layout for the Builder to hash goes as follows:
    // 1) refs_descriptor:bits8 | bits_descriptor:bits8 | data:bitsN
    //
    //  refs_descriptor: ref_count + ((exotic? & 1) * 8) + (mask * 32)
    //                   2 refs (code + data), non-exotic, zero-mask
    //
    //  bits_descriptor: floor(bit_count / 8) + ceil(bit_count, 8)
    //                   floor (5 bits / 8) + ceil(5 bits / 8) = 0 + 1 = 1
    //
    //  data: [0b00110] + [0b100] = [0b00110100] = 0x34 (data + augmented bits)
    //        0b00110 - data (split_depth, special, code, data, Library)
    //        0b100 - augmented bits (Leading 1 + zeroes to make section multiple of eight)
    //
    //  That is: (2 << 16) | (1 << 8) | 0x34 = 131380 for all three.
    //
    // 2) and 3) depth_descriptors: CDEPTH of `code` and CDEPTH of `data`
    // 4) and 5) ref hashes: HASHCU of `code` and HASHCU of `data`

    // Group 1: Computations and arrangements
    s0 PUSH HASHCU // `data` hash
    s2 PUSH HASHCU // `code` hash
    SWAP2
    CDEPTH         // `data` depth
    SWAP
    CDEPTH         // `code` depth
    131380 INT     // (2 << 16) | (1 << 8) | 0x34

    // Group 2: Composition of the Builder
    NEWC
    24 STU  // store refs_descriptor | bits_descriptor | data
    16 STU  // store depth_descriptor for `code`
    16 STU  // store depth_descriptor for `data`
    256 STU // store `code` hash
    256 STU // store `data` hash

    // Group 3: SHA256 hash of the resulting Builder
    ONE HASHEXT_SHA256 // a.k.a. Builder.hash()
}
```

In this form, it consumes 90 **less** gas than its non-`asm` counterpart, while producing identical results.

@anton-trunov @Shvandre @i582 @Gusarich If you agree with this vision, let's remove the `Builder.hash()` and replace the current `contractHash()` with my `asm` implementation :)

novusnota committed:
feat: gas-optimized `asm` implementation of `contractHash()`

And removal of unnecessary `Builder.hash()`, which was originally
intended to be private and only used in previous code of
`contractHash()`

novusnota committed:
Merge branch 'main' into pmakhnev/rewrite-address-related-functions-to-tact

novusnota commented:
Fixing benchmarks, uno momento por favor

novusnota committed:
fix: snapshots

novusnota committed:
Merge branch 'main' into pmakhnev/rewrite-address-related-functions-to-tact
